### PR TITLE
[Parser] support loop op parser

### PR DESCRIPTION
**Pull request by youbeny**
this PR is depend on the previous PR：https://github.com/alibaba/heterogeneity-aware-lowering-and-optimization/pull/215

### CODE REVIEW

File: lib/parser/onnx/onnx_parser.cc [hunk 1]
```diff
@@ -21,6 +21,7 @@
 #include <google/protobuf/text_format.h>
 
 #include <fstream>
+#include <numeric>
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:24]: ok

File: include/halo/lib/ir/onnx_convert.td [hunk 2]
```diff
@@ -159,9 +159,9 @@ def ONNX_LeakyRelu : OpMapping<"LeakyRelu", LeakyRelu> {
 
 def ONNX_Log : OpMapping<"Log", Log>;
 
-def ONNX_Loop : ONNXExtension<"Loop"> {
-  let num_outputs_= 2;
-}
+// def ONNX_Loop : ONNXExtension<"Loop"> {
```
[the remark this answers is not in the document]
Comment (youbeny) [reply include/halo/lib/ir/onnx_convert.td:162]: loop op is special，it is converted in onnx_parser.cc without TableGen

File: lib/parser/tflite/tflite_parser.h [hunk 3]
```diff
@@ -50,7 +50,8 @@ class TFLITEParser : public Parser {
  private:
   void RegisterOp();
   Status ConvertToHaloIR(const tflite::Model& model);
-  Status ConvertOneNode(const tflite::Operator& cur_node,
+  Status ConvertOneNode(std::unique_ptr<IRBuilder>& ir_builder,
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/tflite/tflite_parser.h:53]: you are right

File: lib/parser/tensorflow/tf_parser.h [hunk 4]
```diff
@@ -77,17 +77,21 @@ class TFParser : public Parser {
...
   Constant* CreateConstant(TFAttrs* attrs, DataType data_type,
                            const tensorflow::NodeDef& node_def);
 
 /// create node function auto generatered by tablegen
 #include "tf_convert.h.inc"
 
-  Status ConvertConstNode(const tensorflow::NodeDef& node_def);
-  Status ConvertPlaceholderNode(const tensorflow::NodeDef& node_def);
-  Status ConvertDummyNode(const tensorflow::NodeDef& node_def);
+  Status ConvertConstNode(std::unique_ptr<IRBuilder>& ir_builder,
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/tensorflow/tf_parser.h:89]: ok

File: lib/parser/tensorflow/tf_parser.h [hunk 5]
```diff
@@ -77,17 +77,21 @@ class TFParser : public Parser {
   void Init(BasicBlock* bb, Function* function, const armory::Opts& opts);
   void RegisterOp();
   Status ConvertToHaloIR(const tensorflow::GraphDef& graph_def);
-  Status ConvertOneNode(const tensorflow::NodeDef& cur_node, size_t index);
+  Status ConvertOneNode(std::unique_ptr<IRBuilder>& ir_builder,
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/tensorflow/tf_parser.h:80]: ok

File: lib/parser/onnx/onnx_parser.cc [hunk 6]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
-} // end namespace halo
\ No newline at end of file
+Status ONNXParser::ConvertLoopNode(std::unique_ptr<IRBuilder>& ir_builder,
+                                   const onnx::NodeProto& cur_node) {
+  std::string cur_node_name = cur_node.name();
+  if (cur_node_name.empty()) {
+    // TODO(unknown) current node name must unique
+    cur_node_name = "unkonwn";
+  }
+  std::vector<Def> operands = GetInputOperands(cur_node);
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:725]: ok

File: lib/parser/onnx/onnx_parser.cc [hunk 7]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+    }
+  }
+
+  for (int i = 0, node_size = subgraph.node_size(); i < node_size; ++i) {
+    VLOG(1) << "sub node name: " << subgraph.node(i).name();
+    ConvertOneNode(_ir_builder, subgraph.node(i));
+  }
+
+  // Convert output
+#if USE_DEFINE_OUTPUT
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:767]: in some case, output name defined in subgraph's ```OUTPUT``` proto and loop variable tensor name in subgraph mismatch，which is one of the loop body‘s output nodes，so we do not specify the output node's name in parser

File: lib/parser/onnx/onnx_parser.cc [hunk 8]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+    ConvertConstNode(_c_builder, subgraph.initializer(i));
+  }
+
+  // Convert input
+  auto input_infos_size = subgraph.input_size();
+  for (int i = 0; i < input_infos_size; ++i) {
+    auto name = subgraph.input(i).name();
+    // var_args_name_.emplace(name, cur_node.input(i));
+    auto it = const_input_names.find(name);
+    if (it == const_input_names.end()) {
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:752]: ok

File: lib/parser/onnx/onnx_parser.cc [hunk 9]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+       i < const_inputs_size; ++i) {
+    const_input_names.emplace(subgraph.initializer(i).name());
+    ConvertConstNode(_c_builder, subgraph.initializer(i));
+  }
+
+  // Convert input
+  auto input_infos_size = subgraph.input_size();
+  for (int i = 0; i < input_infos_size; ++i) {
+    auto name = subgraph.input(i).name();
+    // var_args_name_.emplace(name, cur_node.input(i));
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:750]: ok

File: lib/parser/onnx/onnx_parser.cc [hunk 10]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+                                   const onnx::NodeProto& cur_node) {
+  std::string cur_node_name = cur_node.name();
+  if (cur_node_name.empty()) {
+    // TODO(unknown) current node name must unique
+    cur_node_name = "unkonwn";
+  }
+  std::vector<Def> operands = GetInputOperands(cur_node);
+  // first 2 inputs(loop_cnt/loop_cond) is optional
+  for (int i = 0, n = operands.size(); i < n; ++i) {
+    var_args_.push(operands[i].GetType());
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:728]: ok

File: lib/parser/onnx/onnx_parser.cc [hunk 11]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
 }
 
-} // end namespace halo
\ No newline at end of file
+Status ONNXParser::ConvertLoopNode(std::unique_ptr<IRBuilder>& ir_builder,
+                                   const onnx::NodeProto& cur_node) {
+  std::string cur_node_name = cur_node.name();
+  if (cur_node_name.empty()) {
+    // TODO(unknown) current node name must unique
+    cur_node_name = "unkonwn";
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:723]: ?

File: lib/parser/onnx/onnx_parser.cc [hunk 12]
```diff
@@ -676,4 +715,77 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+  // first 2 inputs(loop_cnt/loop_cond) is optional
+  for (int i = 0, n = operands.size(); i < n; ++i) {
+    var_args_.push(operands[i].GetType());
+  }
+
+  ONNXAttrs attrs(cur_node);
+  onnx::GraphProto subgraph;
+  attrs.Process<onnx::GraphProto>("body", &subgraph);
+  auto loop_body = bb_builder_->CreateBasicBlock("bb_" + cur_node_name);
+  // TODO(unkonwn) is there exist memory problem ?
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:735]: no problem, i will remove this TODO comment

File: lib/parser/onnx/onnx_parser.h [hunk 13]
```diff
@@ -19,6 +19,8 @@
 #define HALO_LIB_PARSER_ONNX_ONNXPARSER_H_
 
 #include <functional>
+#include <stack>
+#include <tuple>
```
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.h:23]: The codes use tuple has been deleted，i will remove this

File: lib/parser/onnx/onnx_parser.cc (hunk 11 above)
[the remark this answers is not in the document]
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:723]: ok

File: include/halo/lib/ir/onnx_convert.td (hunk 2 above)
[the remark this answers is not in the document]
Comment (weimingzha0) [reply include/halo/lib/ir/onnx_convert.td:162]: If we don't need it, we can delete it.

File: lib/parser/onnx/onnx_parser.cc [hunk 14]
```diff
@@ -676,4 +711,75 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+  for (int i = 0, const_inputs_size = subgraph.initializer_size();
+       i < const_inputs_size; ++i) {
+    const_input_names.emplace(subgraph.initializer(i).name());
+    ConvertConstNode(_c_builder.get(), subgraph.initializer(i));
+  }
+
+  // Convert input
+  auto input_infos_size = subgraph.input_size();
+  for (int i = 0; i < input_infos_size; ++i) {
+    if (const_input_names.count(subgraph.input(i).name())) {
```
Comment (weimingzha0) [inline lib/parser/onnx/onnx_parser.cc:744]: if (!const_input_names.count(...)) ?

File: lib/parser/onnx/onnx_parser.cc [hunk 15]
```diff
@@ -676,4 +711,75 @@ std::unique_ptr<Parser> CreateONNXParser() {
...
+                                   const onnx::NodeProto& cur_node) {
+  std::string cur_node_name = cur_node.name();
+  if (cur_node_name.empty()) {
+    // TODO(unknown) current node name must unique
+    cur_node_name = "unknown";
+  }
+  const auto& operands = GetInputOperands(cur_node);
+  // first 2 inputs(loop_cnt/loop_cond) is optional
+  for (int i = 0, n = operands.size(); i < n; ++i) {
+    loop_arg_types_.push(operands[i].GetType());
```
Comment (weimingzha0) [inline lib/parser/onnx/onnx_parser.cc:724]: 1. why use stack here? The order will be reversed when you use it in L437.

File: lib/parser/onnx/onnx_parser.cc [hunk 16]
```diff
@@ -404,42 +412,56 @@ Status ONNXParser::ConvertConstNode(const onnx::NodeProto& cur_node) {
...
-                                          Type(data_type, shape));
+  return Type(data_type, shape);
+}
+
+Status ONNXParser::ConvertSubPlaceholderNode(
+    ArgumentBuilder* arg_builder, const onnx::ValueInfoProto& value_info_def) {
+  HLCHECK(!loop_arg_types_.empty());
+  auto& type = loop_arg_types_.top();
+  if (!type.IsValid()) {
+    type = GetType(value_info_def);
```
Comment (weimingzha0) [inline lib/parser/onnx/onnx_parser.cc:439]: why not just use the type in value_info_def?

File: lib/parser/onnx/onnx_parser.cc (hunk 15 above)
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:724]: you are right, the order is wrong, i will reverse it when

File: lib/parser/onnx/onnx_parser.cc (hunk 16 above)
Comment (youbeny) [reply lib/parser/onnx/onnx_parser.cc:439]: value_info_def in loop body maybe invalid:
1.can get data type, shape not define
2.can get shape, data type not define
3.both is not define
 but we can use stack to get the type from loop node

**Review by shuhand — APPROVED** [review]

**Pull request merged**